dsl/flatten.cc

```diff
@@ -0,0 +1,270 @@
+            methods.methods.emplace_back();
+        }
+        return send;
+    }
+
+    unique_ptr<ast::Expression> postTransformSend(core::Context ctx, unique_ptr<ast::Send> send) {
+        if (send->fun == core::Names::sig() || isMethodModifier(*send)) {
+            auto &methods = curMethodSet();
+            ENFORCE(!methods.stack.empty());
+            ENFORCE(methods.methods.size() > methods.stack.back().idx);
+            ENFORCE(methods.methods[methods.stack.back().idx] == nullptr);
+
+            if (isMethodModifier(*send) && send->args.size() >= 1) {
+                auto methodDef = ast::cast_tree<ast::MethodDef>(send->args.front().get());
+                ENFORCE(methodDef);
+                methodDef->setIsSelf(methods.stack.back().isStatic);
+            }
+
+            methods.methods[methods.stack.back().idx] = std::move(send);
+            methods.stack.pop_back();
+
+            return make_unique<ast::EmptyTree>();
+        } else {
+            return send;
+        }
+    }
+
+    unique_ptr<ast::Expression> postTransformClassDef(core::Context ctx, unique_ptr<ast::ClassDef> classDef) {
+        classDef->rhs = addMethods(ctx, std::move(classDef->rhs));
+        return classDef;
+    };
+
+    unique_ptr<ast::Expression> postTransformMethodDef(core::Context ctx, unique_ptr<ast::MethodDef> methodDef) {
+        // if this method is contained in a send like `private` or `protected`, then we should not move it, because
+        // moving the send will do that for us
+        if (skipMethods.contains(methodDef.get())) {
+            return methodDef;
+        }
+        auto &methods = curMethodSet();
+        ENFORCE(!methods.stack.empty());
+        ENFORCE(methods.methods.size() > methods.stack.back().idx);
+        ENFORCE(methods.methods[methods.stack.back().idx] == nullptr);
+
+        // we might need to modify the static-ness of the methods
+        methodDef->setIsSelf(methods.stack.back().isStatic);
+
+        methods.methods[methods.stack.back().idx] = std::move(methodDef);
+        methods.stack.pop_back();
+        return make_unique<ast::EmptyTree>();
+    };
+
+    unique_ptr<ast::Expression> addMethods(core::Context ctx, unique_ptr<ast::Expression> tree) {
+        auto &methods = curMethodSet().methods;
+        if (methods.empty()) {
+            ENFORCE(popCurMethodDefs().empty());
+            return tree;
+        }
+        if (methods.size() == 1 && (ast::cast_tree<ast::EmptyTree>(tree.get()) != nullptr)) {
+            // It was only 1 method to begin with, put it back
+            unique_ptr<ast::Expression> methodDef = std::move(popCurMethodDefs()[0]);
+            return methodDef;
+        }
+
+        auto insSeq = ast::cast_tree<ast::InsSeq>(tree.get());
+        if (insSeq == nullptr) {
+            ast::InsSeq::STATS_store stats;
+            tree = make_unique<ast::InsSeq>(tree->loc, std::move(stats), std::move(tree));
```

Comment:
the previous implementation used to new definitions last, lets keep it that way.
It makes reading code easier.

Reply:
This should put new definitions at the end, and with more recent changes will now only move methods that are actually nested.. (FWIW, this whole section of code was borrowed directly from the old flattening pass with no modifications, because I wanted to keep it as similar as possible.)

Reply:
Speaking of which, does this change mean we have two implementations of flattening in Sorbet?